rs/rosetta-api/icrc1/ledger/src/main.rs

```diff
@@ -135,18 +142,79 @@ fn pre_upgrade() {
         .expect("failed to write instructions consumed to stable memory");
 }
 
+#[cfg(feature = "upgrade-to-memory-manager")]
+#[pre_upgrade]
+fn pre_upgrade() {
+    #[cfg(feature = "canbench-rs")]
+    let _p = canbench_rs::bench_scope("pre_upgrade");
+
+    let start = ic_cdk::api::instruction_counter();
+    UPGRADES_MEMORY.with_borrow_mut(|bs| {
+        Access::with_ledger(|ledger| {
+            let writer = Writer::new(bs, 0);
+            let mut buffered_writer = BufferedWriter::new(8388608, writer);
+            ciborium::ser::into_writer(ledger, &mut buffered_writer)
+                .expect("Failed to write the Ledger state in stable memory");
+            let end = ic_cdk::api::instruction_counter();
+            let instructions_consumed = end - start;
+            let counter_bytes: [u8; 8] = instructions_consumed.to_le_bytes();
+            buffered_writer
+                .write_all(&counter_bytes)
+                .expect("failed to write instructions consumed to UPGRADES_MEMORY");
+        });
+    });
+}
+
 #[post_upgrade]
 fn post_upgrade(args: Option<LedgerArgument>) {
     #[cfg(feature = "canbench-rs")]
     let _p = canbench_rs::bench_scope("post_upgrade");
 
     let start = ic_cdk::api::instruction_counter();
     let mut stable_reader = StableReader::default();
-    LEDGER.with(|cell| {
-        *cell.borrow_mut() = Some(
-            ciborium::de::from_reader(&mut stable_reader).expect("failed to decode ledger state"),
-        );
-    });
+
+    let mut pre_upgrade_instructions_consumed = 0;
+
+    let old_des_result: Result<Ledger<Tokens>, _> = ciborium::de::from_reader(&mut stable_reader);
```

Comment:
We discussed with Islam and Mario about reading the first few bytes and comparing them against the `MemoryManager` [magic bytes](https://github.com/dfinity/stable-structures/blob/817a163e0f8a5a2afaad0469e9434398ac6bf7d3/src/memory_manager.rs#L53), and thereafter deciding how to deserialize the ledger state, but I suppose this way is not worse? At least it is not reading the full serialized state into a  buffer multiple times, or trying to initialize the `MemoryManager` first, so this should be fine.

Reply:
Is it guaranteed that this deserialization fails quickly in case the `MemoryManager` is used? If not, the magic bytes could save some instruction overhead.

Reply:
I don't know if it is guaranteed, but in my tests I saw that failed reads always take ~13.5k instructions while the whole read takes ~30M instructions:
```
2021-05-06 19:17:10.000000024 UTC: [Canister rwlgt-iiaaa-aaaaa-aaaaa-cai] failed read used 13591 instructions, read from memory manager used 25894047 instructions
2021-05-06 19:17:10.000000025 UTC: [Canister rwlgt-iiaaa-aaaaa-aaaaa-cai] failed read used 13589 instructions, read from memory manager used 26058438 instructions
2024-09-02 11:10:13.295420582 UTC: [Canister rwlgt-iiaaa-aaaaa-aaaaa-cai] failed read used 13591 instructions, read from memory manager used 34312326 instructions
2024-09-02 11:10:13.295420634 UTC: [Canister rwlgt-iiaaa-aaaaa-aaaaa-cai] failed read used 13589 instructions, read from memory manager used 36608723 instructions
```

Reply:
I am adding 50 tx between the upgrades, that is why the successful reads increase in instructions.